### ISSUE TITLE
Implement PCGrad (Projecting Conflicting Gradients) for Multi-Task Optimization

### ISSUE THREAD

**Issue by Vectorrent** [0-5788719150923125/praxis#51]
### Background
Currently, our model simultaneously optimizes multiple objectives (main cross-entropy loss and auxiliary routing loss), with a possibility of adding RL-based objectives in the future. However, we're facing challenges with conflicting gradients where improving one task can harm performance on another task.

Research has shown that naively combining losses from multiple tasks can lead to destructive interference between task gradients. This occurs when:

1. Gradients from different tasks conflict (point in opposite directions)
2. Some task gradients dominate others due to magnitude differences
3. The optimization landscape has high curvature

### Proposed Solution
We should implement PCGrad (Projecting Conflicting Gradients) as described in the paper ["Gradient Surgery for Multi-Task Learning"](https://arxiv.org/pdf/2001.06782) by Yu et al. This method identifies when gradients from different tasks conflict and projects them onto the normal plane of each other to remove destructive interference while preserving useful components.

**Research Reference**
Yu, T., Kumar, S., Gupta, A., Levine, S., Hausman, K., & Finn, C. (2020). Gradient Surgery for Multi-Task Learning. NeurIPS 2020. [Paper Link](https://arxiv.org/pdf/2001.06782)

### Implementation Requirements
**Changes to Training Infrastructure**

- Disable PyTorch Lightning's automatic optimization: We would need to call self.automatic_optimization = False in our __init__ method
- Rewrite training step: We'll need to handle gradient computation and application manually

**Implementation Steps**

1. Compute each task loss separately
2. For each task, perform a separate backward pass to obtain task-specific gradients
3. Apply PCGrad to handle conflicts between task gradients
4. Update the model with the modified gradients

### PyTorch Implementation Example
```py
# Disabling automatic optimization in Lightning's __init__
def __init__(self):
    super().__init__()
    self.automatic_optimization = False
    # ... rest of initialization

# In training_step
def training_step(self, batch, batch_idx):
    # Get optimizer
    opt = self.optimizers()
    opt.zero_grad()
    
    # Forward pass
    outputs = self.model(batch)
    
    # Compute losses separately
    ce_loss = self.compute_cross_entropy_loss(outputs, batch["targets"])
    routing_loss = self.compute_routing_loss(outputs, batch["routing_targets"])
    
    # PCGrad implementation
    # 1. Get gradients for main task
    self.manual_backward(ce_loss, retain_graph=True)
    ce_grads = {name: param.grad.clone() for name, param in self.named_parameters() if param.grad is not None}
    
    # Reset gradients
    opt.zero_grad()
    
    # 2. Get gradients for auxiliary task
    self.manual_backward(routing_loss)
    routing_grads = {name: param.grad.clone() for name, param in self.named_parameters() if param.grad is not None}
    
    # 3. Apply PCGrad conflict resolution
    opt.zero_grad()
    
    # Project main task gradient if it conflicts with auxiliary task
    for name, param in self.named_parameters():
        if param.grad is not None and name in ce_grads:
            # Check for conflict
            dot_product = torch.sum(ce_grads[name] * param.grad)
            if dot_product < 0:
                # Project main gradient onto normal plane of auxiliary gradient
                projected_grad = ce_grads[name] - (dot_product / torch.sum(param.grad**2)) * param.grad
                ce_grads[name] = projected_grad
    
    # Project auxiliary task gradient if it conflicts with main task
    opt.zero_grad()
    self.manual_backward(routing_loss)  # Recompute auxiliary gradients
    for name, param in self.named_parameters():
        if param.grad is not None and name in ce_grads:
            # Check for conflict
            dot_product = torch.sum(ce_grads[name] * param.grad)
            if dot_product < 0:
                # Project auxiliary gradient onto normal plane of main gradient
                projected_grad = param.grad - (dot_product / torch.sum(ce_grads[name]**2)) * ce_grads[name]
                param.grad = projected_grad
            else:
                param.grad = param.grad + ce_grads[name]
    
    # 4. Step with modified gradients
    opt.step()
    
    # Log metrics
    self.log("train/ce_loss", ce_loss)
    self.log("train/routing_loss", routing_loss)
    
    # For Lightning, we need to return something, but with manual optimization,
    # returning any of the losses won't affect the optimization
    return ce_loss + routing_loss
```
### Potential Challenges

Computational overhead: PCGrad requires a separate backward pass for each task, potentially doubling the computational cost for two tasks.

- Memory overhead: We need to store gradients for each task separately, which approximately doubles the memory needed for gradients.
- Lightning rewrite: We need to refactor our training loop to use manual optimization in PyTorch Lightning, which requires careful testing.
- Gradient scaling: With multiple tasks having different scales, we may need to normalize gradients to prevent one task from dominating.
- Validation: We'll need to develop a way to validate that PCGrad is actually helping and not hurting performance across tasks.

### Potential Optimizations
If the full PCGrad implementation proves too costly, we could consider:

- Selective application: Apply PCGrad only to shared layers where conflicts matter most
- Periodic application: Apply PCGrad every N steps rather than every step
- Task sampling: For more than two tasks, randomly select task pairs for projection

**Comment by Vectorrent** [comment]
Relevant reading, with some potentially simpler alternatives:
https://medium.com/@baicenxiao/strategies-for-balancing-multiple-loss-functions-in-deep-learning-e1a641e0bcc0

**Comment by Vectorrent** [comment]
We are never going to implement PCGrad. At current scale, that could result in 4x or more the number of forward passes, 4x the memory, 4x the time... it simply isn't worth the effort. As for the other strategies, we've implemented them, and they don't work very well. But naive summation is working fine, and while there might be interference in the gradients - what we are doing is the standard. This isn't something worth tracking anymore.
